### ISSUE TITLE
[Revision] Consolidate and revise .NET threading documentation

### ISSUE THREAD

**Issue by rpetrusha** [dotnet/docs#3819]
# Review/revise threading documentation

## Issues with Existing Topics

The ease of writing asynchronous code had made multi-threaded code very common. Often, though, developers don't fully appreciate the fact that their code may be executing on multiple threads, and do not do not fully appreciate the need to synchronize access to resources. This makes it important to update the documentation on threading, and especially to adequately document synchronization primitives.

At the same time, in migrating documentation from MSDN to docs.microsoft.com, we've managed to create separate threading documentation for C# and Visual Basic -- when only one of the topics (on `lock` for C# and `SyncLock` for Visual Basic) in each documentation set actually concerns a language feature. The bulk of the threading docs are located in the .NET Guide, which are language-independent treatments of managed threading in .NET. 

### C# documentation

1. The C# [Threading (C#](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/threading/) section should be removed after all its content has been migrated to the .NET Framework docs.
2. The URL for each deleted topic should be entered in the [redirection file](https://github.com/dotnet/docs/blob/master/.openpublishing.redirection.json) to redirect it to the version of the topic in the .NET docs.
3. The only C# language content that may need to be preserved is in the [Thread Synchronization](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/threading/thread-synchronization) topic. Possibly, it should be added to the  [lock statement](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/lock-statement).
4. All the other topics have Visual Basic equivalents that are completely or largely identical to the C# topics. They should be moved to the .NET Guide, with both VB and C# code examples appearing in the same topic. 
  
### Visual Basic documentation

1. The Visual Basic [Threading (Visual Basic](https://docs.microsoft.com/en-us/dotnet/visual-basic/programming-guide/concepts/threading/) section should be removed after all its content has been migrated to the .NET Framework docs.
2. The URL for each deleted topic should be entered in the [redirection file](https://github.com/dotnet/docs/blob/master/.openpublishing.redirection.json) to redirect it to the version of the topic in the .NET docs.
3. The only VB language content that may need to be preserved is in the [Thread Synchronization](https://docs.microsoft.com/en-us/dotnet/visual-basic/programming-guide/concepts/threading/thread-synchronization) topic. Possibly, it should be added to the  [SyncLock statement](https://docs.microsoft.com/en-us/dotnet/visual-basic/language-reference/statements/synclock-statement).
4. All the other topics have Visual Basic equivalents that are completely or largely identical to the C# topics. They should be moved to the .NET Guide, with both VB and C# code examples appearing in the same topic. 
  
## .NET Guide

1. We should decide whether to place this section in the .NET Standard guide (which would be my choice) or leave it in the .NET Guide.
2. There should be some consolidation of content. For example, the [Threads and Threading](https://docs.microsoft.com/en-us/dotnet/standard/threading/threads-and-threading) topic includes exception handling best practices; these are best moved to the [Managed Threading Best Practices](https://docs.microsoft.com/en-us/dotnet/standard/threading/managed-threading-best-practices) topic.
3. Each topic in the .NET Guide should be reviewed and, if necessary, revised. The documentation should focus on providing the context for a particular type's use, and on providing good, fairly simple examples (if possible) to illustrate that type or threading concept.
4. We should decide whether the conceptual threading documentation is the appropriate place for a detailed description of threading types, or whether this is best done in the managed reference documentation. In this case, the conceptual docs may need to provide only a brief example and refer developers to the managed reference.
5. The top-level threading portal page should be revised to reflect new content.
6. The TOC should be revised to reflect new content.

## See also

Issue #1117 

//cc @mairaw @BillWagner @JRAlexander @sputier

**Comment by mairaw** [comment]
This is great @rpetrusha. Do you have an idea of when you plan to address this? Or should it go to backlog?

**Comment by pkulikov** [comment]
@rpetrusha we've made some progress on that issue, let's celebrate that 🎆 

> C# documentation
> 1...4

That was completely fixed by #6180 and #7317

> Visual Basic documentation
> 1...4

That was also fixed by #6180. What remains is a possible revision of the [SyncLock Statement](https://docs.microsoft.com/en-us/dotnet/visual-basic/language-reference/statements/synclock-statement) article. 

As for the .NET Guide part...

>1. We should decide whether to place this section in the .NET Standard guide (which would be my choice) or leave it in the .NET Guide.

What is the difference between those two guides? I don't see the .NET Standard guide. There are .NET Guide, .NET Core Guide, and .NET Framework Guide. Is that point still relevant?

> 2. There should be some consolidation of content. For example, the Threads and Threading topic includes exception handling best practices; these are best moved to the Managed Threading Best Practices topic.

That's still true. I think some content might redirected to the API docs, some content is outdated, so might be shortened or removed completely. I'll submit several PRs concerning this point.

> 3. Each topic in the .NET Guide should be reviewed and, if necessary, revised. The documentation should focus on providing the context for a particular type's use, and on providing good, fairly simple examples (if possible) to illustrate that type or threading concept.

Is the last sentence still the way you want to proceed? I think the last sentence would sound better if it started like "The API documentation...". The issue #8121 goes that way: move docs about a particular type's use to the API docs. The conceptual docs provide overview of threading and available .NET types.

> 4. We should decide whether the conceptual threading documentation is the appropriate place for a detailed description of threading types, or whether this is best done in the managed reference documentation. In this case, the conceptual docs may need to provide only a brief example and refer developers to the managed reference.

Hm, exactly what I've meant in my reply to the previous point. So, I might have misunderstood the point 3. #8121 should fix this point, then.

> 5. The top-level threading portal page should be revised to reflect new content.

I think now it references to the docs sections about TPL and PLINQ. Do you foresee more updates required? Or what would be new content?

> 6. The TOC should be revised to reflect new content.

I guess, it would be a side effect of working on the point 2. Usually TOC is updated, whenever articles are updated.

In summary, C# and Visual Basic documentation is cleaned up. The revision of the .NET Guide is in progress. The current WIP is to redirect docs about the synchronization primitives to the API docs, see #8121.
